spring-cloud-gcp-pubsub/src/main/java/org/springframework/cloud/gcp/pubsub/integration/inbound/PubSubInboundChannelAdapter.java

```diff
@@ -153,7 +153,14 @@ public void nack() {
 		}
 		catch (RuntimeException re) {
 			if (this.ackMode == AckMode.AUTO) {
-				message.nack();
+				if (getErrorChannel() != null) {
+					// If the error channel is present, acknowledge the message.
+					// It gets forwarded to the error channel once PubSubException is thrown.
+					message.ack();
+				}
+				else {
+					message.nack();
+				}
 			}
 			throw new PubSubException("Sending Spring message failed.", re);
```

Comment:
Not related, but I would change this anyway:

`"Message producing for "+message+"has failed in: " + this`

Also pay attention how `MessageProducerSupport.sendMessage()` works: if there is a success error sending, there is no an exception re-throwing. So, this your `catch` logic is not going to work, when we have an `errorChannel`.

I would say that we are good with the `	if ((this.ackMode == AckMode.AUTO) || (this.ackMode == AckMode.AUTO_ACK)) {` in the end of this method and the message is going to be acked in normal case and when we have an `errorChannel`. Exactly what we would like to achieve. You just revert all your changes in this method and we are good to go ! 😄

Reply:
Hey Artem, for some reason this catch-block is still actually triggered by the runtime exception.

I see what you are saying with `MessageProducerSupport`, however when I run the code in the debugger, an exception is actually thrown when the message template attempts to send the error message to the channel - [this line](https://github.com/spring-projects/spring-integration/blob/master/spring-integration-core/src/main/java/org/springframework/integration/endpoint/MessageProducerSupport.java#L217).

Any ideas why you think that may be the case? It appears that in that call, the message handler (which threw the exception) somehow gets triggered again. Do we need to implement `ErrorMessageStrategy` or some additional class?

Reply:
So, that is a different story and you have an error during sending error message.
It would be great to see what is an exception there and what are subscribers to this error channel.

On other other hand we really can't guarantee that there is no exceptions during error message sending. 
So, you should consider to move ack logic into `finally` block with some state to determine that it has not been nacked in the `catch` block.

Reply:
Hmm, in that case maybe the error handling setup our stream binder code sample is wrong? I wonder if there is a way to make it so we do not throw an error when sending to `errorChannel`.

The problem is it looks like the error handling invokes the message handler which threw the error initially *again*.

These are the [subscriber config](https://github.com/spring-cloud/spring-cloud-gcp/blob/953f697a9804847d5fce75f8f170200b78e33a97/spring-cloud-gcp-samples/spring-cloud-gcp-pubsub-binder-sample/src/main/java/com/example/SinkExample.java) for the sample that I using to test error handling.

And the stack trace below - the first two logged lines by `com.example.SinkExample` is logged by my error handler.

Then the line logged by `o.s.integration.handler.LoggingHandler` is the error that I get in the `catch` block. This is what threw the error during error message sending, causing me to enter the catch block. It looks like the error that get thrown during error handling is the same as the initial error?

stack trace:

```
2019-10-02 16:48:43.514 ERROR 18967 --- [sub-subscriber1] com.example.SinkExample                  : The message that was sent is now processed by the error handler.
2019-10-02 16:48:43.516 ERROR 18967 --- [sub-subscriber1] com.example.SinkExample                  : Failed message: GenericMessage [payload=byte[97], headers={id=45f2dc80-6b45-955b-4806-17725f4a9701, contentType=application/json, gcp_pubsub_original_message=PushedAcknowledgeablePubsubMessage{projectId='cowzow', subscriptionName='anonymous.input-topic.29f225e0-67a5-48ff-827f-8363e524fb58', message=data: "{\"body\":\"dfasdf\",\"username\":\"medeadlift\",\"createdAt\":\"2019-10-02T16:47:27.047\",\"throwError\":true}"
attributes {
  key: "contentType"
  value: "application/json"
}
message_id: "437695676167497"
publish_time {
  seconds: 1570049249
  nanos: 519000000
}
}, timestamp=1570049321197}]

2019-10-02 16:48:43.517 ERROR 18967 --- [sub-subscriber1] o.s.integration.handler.LoggingHandler   : org.springframework.messaging.MessagingException: Exception thrown while invoking SinkExample#handleMessage[1 args]; nested exception is java.lang.RuntimeException: An error was triggered in the message handler!, failedMessage=GenericMessage [payload=byte[97], headers={id=45f2dc80-6b45-955b-4806-17725f4a9701, contentType=application/json, gcp_pubsub_original_message=PushedAcknowledgeablePubsubMessage{projectId='cowzow', subscriptionName='anonymous.input-topic.29f225e0-67a5-48ff-827f-8363e524fb58', message=data: "{\"body\":\"dfasdf\",\"username\":\"medeadlift\",\"createdAt\":\"2019-10-02T16:47:27.047\",\"throwError\":true}"
attributes {
  key: "contentType"
  value: "application/json"
}
message_id: "437695676167497"
publish_time {
  seconds: 1570049249
  nanos: 519000000
}
}, timestamp=1570049321197}]
	at org.springframework.cloud.stream.binding.StreamListenerMessageHandler.handleRequestMessage(StreamListenerMessageHandler.java:64)
	at org.springframework.integration.handler.AbstractReplyProducingMessageHandler.handleMessageInternal(AbstractReplyProducingMessageHandler.java:127)
	at org.springframework.integration.handler.AbstractMessageHandler.handleMessage(AbstractMessageHandler.java:177)
	at org.springframework.integration.dispatcher.AbstractDispatcher.tryOptimizedDispatch(AbstractDispatcher.java:115)
	at org.springframework.integration.dispatcher.UnicastingDispatcher.doDispatch(UnicastingDispatcher.java:133)
	at org.springframework.integration.dispatcher.UnicastingDispatcher.dispatch(UnicastingDispatcher.java:106)
	at org.springframework.integration.channel.AbstractSubscribableChannel.doSend(AbstractSubscribableChannel.java:73)
	at org.springframework.integration.channel.AbstractMessageChannel.send(AbstractMessageChannel.java:453)
	at org.springframework.integration.channel.AbstractMessageChannel.send(AbstractMessageChannel.java:403)
	at org.springframework.messaging.core.GenericMessagingTemplate.doSend(GenericMessagingTemplate.java:187)
	at org.springframework.messaging.core.GenericMessagingTemplate.doSend(GenericMessagingTemplate.java:166)
	at org.springframework.messaging.core.GenericMessagingTemplate.doSend(GenericMessagingTemplate.java:47)
	at org.springframework.messaging.core.AbstractMessageSendingTemplate.send(AbstractMessageSendingTemplate.java:109)
	at org.springframework.integration.endpoint.MessageProducerSupport.sendMessage(MessageProducerSupport.java:198)
	at org.springframework.cloud.gcp.pubsub.integration.inbound.PubSubInboundChannelAdapter.consumeMessage(PubSubInboundChannelAdapter.java:151)
	at org.springframework.cloud.gcp.pubsub.core.subscriber.PubSubSubscriberTemplate.lambda$subscribeAndConvert$1(PubSubSubscriberTemplate.java:154)
	at com.google.cloud.pubsub.v1.MessageDispatcher$4.run(MessageDispatcher.java:379)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:295)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.RuntimeException: An error was triggered in the message handler!
	at com.example.SinkExample.handleMessage(SinkExample.java:41)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springframework.messaging.handler.invocation.InvocableHandlerMethod.doInvoke(InvocableHandlerMethod.java:171)
	at org.springframework.messaging.handler.invocation.InvocableHandlerMethod.invoke(InvocableHandlerMethod.java:120)
	at org.springframework.cloud.stream.binding.StreamListenerMessageHandler.handleRequestMessage(StreamListenerMessageHandler.java:55)
	... 23 more
```

Reply:
i don't see an *exception* in the log you show. That is just an `ERROR` logging message.
It fully doesn't show us that your `RuntimeException` is re-thrown after sending into the `errorChannel`.

Please, check an exception in debug mode when it is thrown during sending into the `errorChannel`.
I mean place a break point in the `MessageProducerSupport.sendErrorMessageIfNecessary()` and look for an exception which is re-thrown from the `this.messagingTemplate.send(channel, buildErrorMessage(message, exception));`

Reply:
It might be the fact that you need to add a subscriber to the *destination specific error channel*, not a global one with that `errorChannel` bean name: https://cloud.spring.io/spring-cloud-static/spring-cloud-stream/2.2.1.RELEASE/spring-cloud-stream.html#spring-cloud-stream-overview-error-handling. 

I guess if you don't have one, there is something like *re-throwing subscriber* for the global `errorChannel`